src/main/scala/org/broadinstitute/hail/driver/ExportVCF.scala

```diff
@@ -205,7 +205,58 @@ object ExportVCF extends Command {
           sb.append("GT:AD:DP:GQ:PL")
         gs.foreach { g =>
           sb += '\t'
-          sb.append(g)
+
+          sb.append(g.gt.map { gt =>
+            val p = Genotype.gtPair(gt)
+            s"${ p.j }/${ p.k }"
+          }.getOrElse("./."))
+
+          val missingField: Array[String] = Array(".")
+
+          def concatAppend(toAppend: Array[_]*): Unit = {
+            sb += ':'
+            toAppend.foreachBetween(_.foreachBetween(sb.append(_))(sb += ','))(sb += ':')
+          }
+
+          (g.ad,
+            g.dp.map(Array(_)),
+            g.gq.map(Array(_)),
+            if (g.isDosage)
+              g.dosage
+            else
+              g.pl) match {
+            case (None, None, None, None) =>
+            case (Some(ad), None, None, None) =>
+              concatAppend(ad)
+            case (None, Some(dp), None, None) =>
+              concatAppend(missingField, dp)
+            case (Some(ad), Some(dp), None, None) =>
+              concatAppend(ad, dp)
+            case (None, None, Some(gq), None) =>
+              concatAppend(missingField, missingField, gq)
+            case (Some(ad), None, Some(gq), None) =>
+              concatAppend(ad, missingField, gq)
+            case (None, Some(dp), Some(gq), None) =>
+              concatAppend(missingField, dp, gq)
+            case (Some(ad), Some(dp), Some(gq), None) =>
+              concatAppend(ad, dp, gq)
+            case (None, None, None, Some(plOrD)) =>
+              concatAppend(missingField, missingField, missingField, plOrD)
+            case (Some(ad), None, None, Some(plOrD)) =>
+              concatAppend(ad, missingField, missingField, plOrD)
+            case (None, Some(dp), None, Some(plOrD)) =>
+              concatAppend(missingField, dp, missingField, plOrD)
+            case (Some(ad), Some(dp), None, Some(plOrD)) =>
+              concatAppend(ad, dp, missingField, plOrD)
+            case (None, None, Some(gq), Some(plOrD)) =>
+              concatAppend(missingField, missingField, gq, plOrD)
+            case (Some(ad), None, Some(gq), Some(plOrD)) =>
+              concatAppend(ad, missingField, gq, plOrD)
+            case (None, Some(dp), Some(gq), Some(plOrD)) =>
+              concatAppend(missingField, dp, gq, plOrD)
+            case (Some(ad), Some(dp), Some(gq), Some(plOrD)) =>
+              concatAppend(ad, dp, gq, plOrD)
+          }
```

Comment:
Sorry, my example was too terse.  I think there only needs to be 4 cases.  You just need to determine the number of fields you're going to output.  Then you can have helper functions based on type (Array[Int], Option[Int], Option[Array[Int]], Array[Double] are all you need, I think.)

Your current version still does quite a bit of cons'ing: the argument to concat append is an allocated array, it is an array of Any, so it boxes, e.g., integers, and `sb.append` on an Any argument calls `toString` so you're allocating all the intermediate stings.